src/toolbars/AudioSetupToolBar.cpp

```diff
@@ -0,0 +1,840 @@
+
+*******************************************************************//*!
+
+\class AudioSetupToolBar
+\brief A toolbar to allow easier changing of input and output devices .
+
+*//*******************************************************************/
+
+
+
+#include "AudioSetupToolBar.h"
+#include "ToolManager.h"
+
+#include <thread>
+
+// For compilers that support precompilation, includes "wx/wx.h".
+#include <wx/wxprec.h>
+
+#include <wx/setup.h> // for wxUSE_* macros
+
+#ifndef WX_PRECOMP
+#include <wx/app.h>
+#include <wx/choice.h>
+#include <wx/event.h>
+#include <wx/intl.h>
+#include <wx/menu.h>
+#include <wx/menuitem.h>
+#include <wx/settings.h>
+#include <wx/sizer.h>
+#include <wx/statbmp.h>
+#include <wx/stattext.h>
+#include <wx/string.h>
+#include <wx/tooltip.h>
+#endif
+
+#include "../ActiveProject.h"
+#include "../TrackPanel.h"
+
+#include "AColor.h"
+#include "AllThemeResources.h"
+#include "AudioIOBase.h"
+#include "ImageManipulation.h"
+#include "DeviceToolBar.h"
+#include "../KeyboardCapture.h"
+#include "Prefs.h"
+#include "Project.h"
+#include "../ProjectWindows.h"
+#include "../ShuttleGui.h"
+#include "../widgets/Grabber.h"
+#include "DeviceManager.h"
+#include "../prefs/PrefsDialog.h"
+#include "../prefs/DevicePrefs.h"
+#include "../widgets/AButton.h"
+#include "../widgets/AudacityMessageBox.h"
+#include "../widgets/BasicMenu.h"
+#include "../widgets/Grabber.h"
+#include "../widgets/wxWidgetsWindowPlacement.h"
+
+#if wxUSE_ACCESSIBILITY
+#include "../widgets/WindowAccessible.h"
+#endif
+
+namespace {
+   static constexpr int kHost = 15000;
+   static constexpr int kInput = 15200;
+   static constexpr int kInputChannels = 15400;
+   static constexpr int kOutput = 15600;
+   static constexpr int kAudioSettings = 15800;
+
+   class ViewDeviceSettingsDialog final : public PrefsDialog
+   {
+   public:
+      ViewDeviceSettingsDialog(wxWindow* parent, AudacityProject& project,
+         const TranslatableString& title, PrefsPanel::Factories& factories,
+         int page)
+         : PrefsDialog(parent, &project, title, factories)
+         , mPage(page)
+      {
+      }
+
+      long GetPreferredPage() override
+      {
+         return mPage;
+      }
+
+      void SavePreferredPage() override
+      {
+      }
+
+   private:
+      const int mPage;
+   };
+}
+
+IMPLEMENT_CLASS(AudioSetupToolBar, ToolBar);
+
+////////////////////////////////////////////////////////////
+/// Methods for AudioSetupToolBar
+////////////////////////////////////////////////////////////
+
+BEGIN_EVENT_TABLE(AudioSetupToolBar, ToolBar)
+   EVT_BUTTON(ID_AUDIO_SETUP_BUTTON, AudioSetupToolBar::OnAudioSetup)
+END_EVENT_TABLE()
+
+//Standard constructor
+AudioSetupToolBar::AudioSetupToolBar( AudacityProject &project )
+: ToolBar( project, AudioSetupBarID, XO("Audio Setup"), wxT("Audio Setup") )
+{
+   mSubscription = DeviceManager::Instance()->Subscribe(
+      *this, &AudioSetupToolBar::OnRescannedDevices );
+
+   mSizer = nullptr;
```

Comment:
Don't assign, when you can use a member initializer.
But don't use a member initializer, when you can use an in-class initializer!